Assets/MRTK/SDK/Experimental/Drawstring/Scripts/ElasticSystem.cs

```diff
@@ -0,0 +1,196 @@
+
+        public ElasticSystem(T initialValue, T initialVelocity,
+                                ElasticExtentProperties<T> extentInfo, ElasticProperties elasticProperties)
+        {
+            this.extentInfo = extentInfo;
+            this.elasticProperties = elasticProperties;
+            currentValue = initialValue;
+            currentVelocity = initialVelocity;
+        }
+
+        /// <summary>
+        /// Update the internal state of the damped harmonic oscillator, given the forcing/desired value.
+        /// </summary>
+        /// <param name="forcingValue">Input value, for example, a desired manipulation position.</param>
+        /// <param name="deltaTime">Amount of time that has passed since the last update.</param>
+        public abstract T ComputeIteration(T forcingValue, float deltaTime);
+
+        /// <summary>
+        /// Query the elastic system for the current instantaneous value
+        /// </summary>
+        /// <returns>Current value of the elastic system</returns>
+        public T GetCurrentValue() => currentValue;
+    }
+
+    internal class LinearElasticSystem : ElasticSystem<float>
+    {
+        public LinearElasticSystem(float initialValue, float initialVelocity,
+                                ElasticExtentProperties<float> extentInfo, ElasticProperties elasticProperties)
+        : base(initialValue, initialVelocity, extentInfo, elasticProperties) { }
+
+
+        /// <summary>
+        /// Update the internal state of the damped harmonic oscillator, given the forcing/desired value.
+        /// </summary>
+        /// <param name="forcingValue">Input value, for example, a desired manipulation position.</param>
+        /// <param name="deltaTime">Amount of time that has passed since the last update.</param>
+        public override float ComputeIteration(float forcingValue, float deltaTime)
+        {
+            // F = -kx - (drag * v)
+            var force = (forcingValue - currentValue) * elasticProperties.hand_k - elasticProperties.drag * currentVelocity;
+
+            // Distance that the current stretch value is from the end limit.
+            float distFromEnd = extentInfo.maxStretch - currentValue;
+
+            // If we are extended beyond the end cap,
+            // add one-sided force back to the center.
+            if (currentValue > extentInfo.maxStretch)
+            {
+                force += distFromEnd * elasticProperties.end_k;
+            }
+            else
+            {
+                // Otherwise, add standard bidirectional magnetic/snapping force towards the end marker. (optional)
+                if (extentInfo.snapToMax)
+                {
+                    force += (distFromEnd) * elasticProperties.end_k * (1.0f - Mathf.Clamp01(Mathf.Abs(distFromEnd / elasticProperties.snap_radius)));
+                }
+            }
+
+            distFromEnd = extentInfo.minStretch - currentValue;
+            if (currentValue < extentInfo.minStretch)
```

Comment:
Since you do the same code for min and max, it might be worth having this be a helper function so that you can avoid both copy pasta and bugs in the future sneaking in because of fixing in one location and not the other.

Reply:
Done, fixed with local functions/closures.